Update Plane2D step after resetDimensions

yann-lty opened the pull request:
Plane2D's step is directly related to effective width and needs to be updated when dimension changes.
This fixes an issue when mixing images with different resolutions that can result in wrong buffer accesses, as shown here:
![popsift_plane2D](https://user-images.githubusercontent.com/1674646/55020699-a8432580-4ff7-11e9-8ba8-39957a0f2008.jpg)
Re-using a previously allocated buffer (first figure) from the pool to work on another image with different width/height leads to erroneous results (second figure); updating the step fix this (third image).
Probably related to #56.

yann-lty commented:
This can not be merged as is; we need to know the **real** buffer size in Plane2D to be able to check if `resetDimensions` to something larger than current effective size can be done.
At the moment, the sanity check `if( w*sizeof(T) > this->getPitch() ) {` is broken.

griwodz commented:
If you change step in Plane2D<T>, you must do it in compliance with Texture alignment rules. Since this is a card-specific value, you are not supposed to guess. So "this->step = w * this->elemSize();" is dangerous just because of that.

If you dare doing this, you should round up to the next multiple of 256 (there are cards that need 32, most need 128, but a few new 256).

griwodz commented:
The new commit adds a size_t _total_bytes to PlaneBase to hold the allocated memory. That is somewhat hackish because PlaneBase objects are instantiated on the GPU side, but this _total_bytes will only be initialized correctly on the CPU side.
The resetDimension test in s_image has been changed to ->canResetDimensions(), which uses a conservative heuristic for pitch estimation. This is not fully safe, but NVidia has published neither documentation nor API for safe pitch computation.

yann-lty commented:
I've made some tests but this does not to solve the issue of buffer access when a Plane2D was "resized".
![image](https://user-images.githubusercontent.com/1674646/55638812-5f922600-57c8-11e9-8931-bfaa824a0149.png)
I think that at some point (still have to find where), we are using step instead of useful width and it leads to incorrect index computation/buffer access. This is probably why my (hackish) first commit seemed to fix the problem.

yann-lty committed:
[plane_2d] update step for host buffers after resetDimensions

Host (CPU) Plane2D references contiguous memory and doesn't have pitch. Its step must be updated when it changes dimensions.
* split resetDimensions into resetDimensionsHost and resetDimensionsDev
  * resetDimensionsHost: reference contiguous memory => update Plane2D's step to width * elemSize()
  * resetDimensionsDev: reference pitched memory => let the step untouched

yann-lty committed:
[plane_2D] rename step to _pitchInBytes

fabiencastan reviewed (approved):
LGTM

yann-lty commented:
The problem was not about updating the step of device Plane2Ds, but host ones.
We use plain memcpy to load the image on the CPU. If host Plane2D's step is not updated, the memcpyToDevice, which relies on it, will not copy the buffer to the device properly.
* Host: contiguous memory without pitch, must always remains width * elemSize() => should be updated by resetDimensions
* Device: pitched GPU memory, step must be set at allocation and never be modified (at least until re-allocation) => should NOT be updated by resetDimensions

Introducing resetDimensionsDev and resetDimensionsHost to handle this.
Also, made "step" a protected member and renamed it to _pitchInBytes to simplify naming (I can rollback this if it's not wanted).

simogasp commented on src/popsift/common/plane_2d.h:
```diff
@@ -285,10 +288,18 @@ template <typename T> class Plane2D : public PitchPlane2D<T>
+     *  @warning: pitch is not updated (device side)
      */
-    __host__ void resetDimensions( int w, int h );
+    __host__ void resetDimensionsDev( int w, int h );
```
`resetDimensionsDevice` it's more clear to me

yann-lty replied:
Agree, but would require to change existing functions (allocDev, freeDev...) for consistency.

fabiencastan replied:
Yes, this convention is used in many places. If changed, it should be in another PR.

simogasp replied:
ok i didn't see the others.

fabiencastan commented:
This looks fine to me now.
@griwodz Can you review and merge?
